x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -853,13 +896,43 @@ private static LogicalPlan maybePushDownPastUnary(Filter filter, UnaryPlan unary
      *
      * Ordering the evals before the orderBys has the advantage that it's always possible to order the plans like this.
      * E.g., in the example above it would not be possible to put the eval after the two orderBys.
+     *
+     * In case one of the eval's fields would shadow the orderBy's attributes, we rename the attribute first.
+     *
+     * E.g.
+     *
+     * ... | sort a | eval a = b + 1 | ...
+     *
+     * becomes
+     *
+     * ... | eval $$a = a | eval a = b + 1 | sort $$a | drop $$a
      */
     protected static class PushDownEval extends OptimizerRules.OptimizerRule<Eval> {
         @Override
         protected LogicalPlan rule(Eval eval) {
             LogicalPlan child = eval.child();
 
             if (child instanceof OrderBy orderBy) {
+                Set<String> evalFieldNames = eval.fields().stream().map(NamedExpression::name).collect(Collectors.toSet());
+
+                // Look for attributes in the OrderBy's expressions and create aliases with temporary names for them.
+                AttributeReplacement nonShadowedOrders = renameAttributesInExpressions(evalFieldNames, orderBy.order());
+
+                AttributeMap<Alias> aliasesForShadowedOrderByAttrs = nonShadowedOrders.replacedAttributes;
+                @SuppressWarnings("unchecked")
+                List<Order> newOrder = (List<Order>) (Object) nonShadowedOrders.rewrittenExpressions;
+
+                if (aliasesForShadowedOrderByAttrs.isEmpty() == false) {
+                    List<Alias> newAliases = aliasesForShadowedOrderByAttrs.values().stream().toList();
+
+                    LogicalPlan plan = new Eval(Source.EMPTY, orderBy.child(), newAliases);
```

Comment:
Keep the current node source.

Reply:
I think using the source of the `OrderBy` is the best option - here, we're renaming some attributes for it. (Current node source contains the already overwritten/shadowed attributes, instead.)